File: cpp/456. [M] 132 Pattern.cpp
// https://leetcode.com/problems/two-sum/

#include <vector>
#include <unordered_map>
#include <assert.h>
#include <iostream>
#include <stack>

using namespace std;;

class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        stack<pair<int, int>> s;
        for (int n : nums) {
            if (s.empty() || n < s.top().first) {
                s.push({n, n});
                continue;
            }

            if (n == s.top().first)
                continue;

            int bottom = s.top().first;
            while (!s.empty() && n >= s.top().second) {
                s.pop();
            }

            if (!s.empty() && s.top().first < n)
                return true;

            s.push({bottom, n});
        }
        return false;
    }
};

int main() {
    Solution sol = Solution();
    vector<int> input = {1,2,3,4,5,6,7};
}

File: python/456. [M] 132 Pattern.py
from typing import List


class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        n = len(nums)
        if n < 3:
            return False
        left = nums[0]
        for i in range(1, n - 1):
            mid = nums[i]
            if mid <= left:
                left = mid
                continue
            for right in nums[i+1:]:
                if left < right < mid:
                    return True
            
        return False
        

if __name__ == '__main__':
    sol = Solution()
    print(sol.find132pattern([1,2,3]))
    print(sol.find132pattern([1,2,3,2]))
    print(sol.find132pattern([1,4,3]))
